src/libsyntax/parse/parser.rs

```diff
@@ -2211,6 +2211,11 @@ impl<'a> Parser<'a> {
                         ex = ExprBreak(None);
                     }
                     hi = self.last_span.hi;
+                } else if self.token.is_keyword(keywords::Let) {
+                    // Catch this syntax error here, instead of in `check_strict_keywords`, so
+                    // that we can explicitly mention that let is not to be used as an expression
+                    let msg = "`let` is not an expression, so it cannot be used in this way";
```

Comment:
This error message needs some work: why single `let` out here? Rather than other keywords, it seems oddly specific. You probably mean that "a `let` statement is not an expression", not that `let ` it self is not. It is not clear what "this way" means.

Maybe better, something like: `expected expression, found statement; variable declaration using `let` is a statement`. Perhaps the bit after the semicolon should be a note.

Note that neither then before or after message fit the `let let x = ...` case, where we expect an identifier, not a keyword, so the original error is better, and the new one (either as well as or instead of) is wrong.